scapy/layers/tls/automaton_cli.py

```diff
@@ -0,0 +1,944 @@
+        pass
+
+    @ATMT.condition(SSLv2_WAITING_CLIENTDATA, prio=1)
+    def sslv2_add_ClientData(self):
+        if not self.data_to_send:
+            data = raw_input()
+            data = data.replace("\\r", "\r")
+            data = data.replace("\\n", "\n")
+        else:
+            data = self.data_to_send.pop()
+            self.vprint("> Read from list: %s" % data)
+        if data == "quit":
+            return
+        if self.linebreak:
+            data += "\n"
+        self.add_record(is_sslv2=True)
+        self.add_msg(Raw(data))
+        raise self.SSLv2_ADDED_CLIENTDATA()
+
+    @ATMT.condition(SSLv2_WAITING_CLIENTDATA, prio=2)
+    def sslv2_no_more_ClientData(self):
+        raise self.SSLv2_CLOSE_NOTIFY()
+
+    @ATMT.state()
+    def SSLv2_ADDED_CLIENTDATA(self):
+        pass
+
+    @ATMT.condition(SSLv2_ADDED_CLIENTDATA)
+    def sslv2_should_send_ClientData(self):
+        self.flush_records()
+        raise self.SSLv2_SENT_CLIENTDATA()
+
+    @ATMT.state()
+    def SSLv2_SENT_CLIENTDATA(self):
+        raise self.SSLv2_WAITING_SERVERDATA()
+
+    @ATMT.state()
+    def SSLv2_WAITING_SERVERDATA(self):
+        self.get_next_msg(0.3, 1)
+        raise self.SSLv2_RECEIVED_SERVERDATA()
+
+    @ATMT.state()
+    def SSLv2_RECEIVED_SERVERDATA(self):
+        pass
+
+    @ATMT.condition(SSLv2_RECEIVED_SERVERDATA)
+    def sslv2_should_handle_ServerData(self):
+        if not self.buffer_in:
+            raise self.SSLv2_WAITING_CLIENTDATA()
+        p = self.buffer_in[0]
+        print "> Received: %s" % p.load
+        if p.load.startswith("goodbye"):
+            raise self.SSLv2_CLOSE_NOTIFY()
+        self.buffer_in = self.buffer_in[1:]
+        raise self.SSLv2_HANDLED_SERVERDATA()
+
+    @ATMT.state()
+    def SSLv2_HANDLED_SERVERDATA(self):
+        raise self.SSLv2_WAITING_CLIENTDATA()
+
+    @ATMT.state()
+    def SSLv2_CLOSE_NOTIFY(self):
+        """
+        There is no proper way to end an SSLv2 session.
+        We try and send a 'goodbye' message as a substitute.
+        """
+        self.vprint()
+        self.vprint("Trying to send a 'goodbye' to the server...")
+
+    @ATMT.condition(SSLv2_CLOSE_NOTIFY)
+    def sslv2_close_session(self):
+        self.add_record()
+        self.add_msg(Raw('goodbye'))
+        try:
+            self.flush_records()
+        except:
+            self.vprint("Could not send our goodbye. The server probably stopped.")
+        self.socket.close()
+        raise self.FINAL()
+
+    ######################### TLS 1.3 handshake ###############################
+
+    @ATMT.state()
+    def TLS13_START(self):
+        pass
+
+    @ATMT.condition(TLS13_START)
+    def tls13_should_add_ClientHello(self):
+        # we have to use the legacy, plaintext TLS record here
+        self.add_record(is_tls13=False)
+        if self.client_hello:
+            p = self.client_hello
+        else:
+            # uncomment these lines (and provide the right IP) for testing TLS 1.3
```

Comment:
This is mostly a helper which would deserve its own example file/documentation somewhere else... For now I rewrote the commented lines to make it more domain-agnostic.